Replace deprecated binder2nd and binder1st

phcerdan opened the pull request:
binder2nd generates error in c++17

Replace by lambdas when possible and with std::function when the type is
required.

decltype(lambda) does not work, so we bind the lambda to a std::function.

Other option also used is std::bind and std::placeholders.

FIX #1221


# Checklist

- [NA] Unit-test of your feature with [Catch](http://dgtal.org/doc/stable/moduleCatch.html).
- [NA] Doxygen documentation of the code completed (classes, methods, types, members...)
- [NA] Documentation module page added or updated.
- [x] New entry in the [ChangeLog.md](https://github.com/DGtal-team/DGtal/blob/master/ChangeLog.md) added.
- [ ] No warning raised in Debug ```cmake``` mode (otherwise, Travis C.I. will fail).
- [x] All continuous integration tests pass (Travis & appveyor)

rolanddenis commented:
Thanks !

In fact, we can make a `decltype` on an *already-existing* lambda but compilation will fail due to the Functor concept in DGtal that except a functor to be assignable. Since a lambda is not assignable, I agree that using `std::function` is the best way !

There is still one `std::binder1st` left in `DGtal/graph/DistanceBreadthFirstVisitor.h` :octocat:

rolanddenis commented on tests/kernel/testImagesSetsUtilities.cpp:
```diff
@@ -203,9 +203,11 @@ bool testSetFromImage()
+  using ValuePredicate = std::function<bool(Image::Value)>;
+  ValuePredicate equalTo1 = [](Image::Value&& v)
+    {
+      return std::forward<Image::Value>(v) == 1;
```
You can also simply pass `v` by const-ref and test `v == 1`. In fact, using `&&` here forbids lvalue parameters, like in the code
```
const auto fn = [] ( int&& v ) { return v == 1; };
int a = 3;
bool dummy = fn(a);
```
Here, since the `std::function` signature specifies a pass-by-value parameter, it doesn't fail.

Therefore, I also suggest to use a `std::function<bool(Image::Value const&)>`.

phcerdan replied:
Ahh, makes sense, still rusty with std::forward usage. I will change it in 12 hours

phcerdan commented:
Good points, I will polish it tomorrow morning, thanks!

dcoeurjo commented:
(tagged as 0.9.4)

phcerdan commented:
I have changed the signature of std::function as @rolanddenis suggested.

Replaced binder1st with std::function, and its instantiation, using std::bind1st, with a std::bind and std::placeholder.

rolanddenis commented:
Thanks !
I have also removed useless `boost/bind.hpp` includes.

dcoeurjo commented:
thanks @phcerdan & @rolanddenis 

Merging..
